src/core/server/saved_objects/migrationsv2/index.ts

```diff
@@ -0,0 +1,857 @@
+    if (Either.isRight(res)) {
+      stateP = {
+        ...stateP,
+        controlState: 'LEGACY_REINDEX_WAIT_FOR_TASK',
+        legacyReindexTaskId: res.right.taskId,
+      };
+    }
+    return stateP;
+  } else if (stateP.controlState === 'LEGACY_REINDEX_WAIT_FOR_TASK') {
+    const res = resW as ResponseType<typeof stateP.controlState>;
+    if (Either.isRight(res)) {
+      return {
+        ...stateP,
+        controlState: 'LEGACY_DELETE',
+      };
+    } else {
+      if (
+        (res.left.type === 'index_not_found_exception' && res.left.index === stateP.legacy) ||
+        res.left.type === 'target_index_had_write_block'
+      ) {
+        // index_not_found_exception for the legacy index, another instance
+        // already complete the LEGACY_DELETE step.
+        //
+        // target_index_had_write_block: another instance already completed the
+        // SET_SOURCE_WRITE_BLOCK step.
+        //
+        // If we detect that another instance has already completed a step, we
+        // can technically skip ahead in the process until after the completed
+        // step. However, by not skipping ahead we limit branches in the
+        // control state progression and simplify the implementation.
+        return { ...stateP, controlState: 'LEGACY_DELETE' };
+      } else if (
+        res.left.type === 'index_not_found_exception' &&
+        res.left.index === stateP.source.value
+      ) {
+        return {
+          ...stateP,
+          controlState: 'FATAL',
+          logs: [
+            ...stateP.logs,
+            {
+              level: 'error',
+              message: `LEGACY_REINDEX failed because the reindex destination index [${stateP.source.value}] does not exist.`,
+            },
+          ],
+        };
+      }
+      return { ...stateP, controlState: 'FATAL' };
+    }
+  } else if (stateP.controlState === 'LEGACY_DELETE') {
+    const res = resW as ResponseType<typeof stateP.controlState>;
+    if (Either.isLeft(res)) {
+      if (
+        res.left.type === 'remove_index_not_a_concrete_index' ||
+        (res.left.type === 'index_not_found_exception' && res.left.index === stateP.legacy)
+      ) {
+        // index_not_found_exception, another Kibana instance already
+        // deleted the legacy index
+        //
+        // remove_index_not_a_concrete_index, another Kibana instance already
+        // deleted the legacy index and created a .kibana alias
+        //
+        // If we detect that another instance has already completed a step, we
+        // can technically skip ahead in the process until after the completed
+        // step. However, by not skipping ahead we limit branches in the
+        // control state progression and simplify the implementation.
+        return { ...stateP, controlState: 'SET_SOURCE_WRITE_BLOCK' };
+      } else if (
+        res.left.type === 'index_not_found_exception' &&
+        res.left.index === stateP.source.value
```

Comment:
Yes, but Kibana shouldn't ever delete the intermediate index, it will only delete the legacy index e.g. the case above this one. I'll add a comment :+1: